pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/QualifiableNode.java

```diff
@@ -0,0 +1,173 @@
+        public static QualifiedName makeClassOf(QualifiedName parent, String className) {
+            QualifiedName qname = new QualifiedName();
+            qname.packages = parent.packages;
+            if (parent.classes[0] != null) {
+                qname.classes = Arrays.copyOf(parent.classes, parent.classes.length + 1);
+                qname.classes[parent.classes.length] = className;
+            } else {
+                qname.classes[0] = className;
+            }
+
+            return qname;
+        }
+
+        // Might be useful with type resolution
+        public static QualifiedName parseCanonicalName(String canon) {
+            throw new UnsupportedOperationException();
+        }
+
+        /** Returns a normalized method name (not Java-canonical!) */
+        private static String getOperationName(String methodName, String[] paramTypes) {
+            StringBuilder sb = new StringBuilder();
+            sb.append(methodName);
+            sb.append('(');
+            int last = paramTypes.length - 1;
+            for (int i = 0; i < last; i++) {
+                sb.append(paramTypes[i]);
+                sb.append(',');
+            }
+
+            if (last > -1) {
+                sb.append(paramTypes[last]);
+            }
+
+            sb.append(')');
+
+            return sb.toString();
+        }
+
+        /** Sets the class to the specified name, truncates the array to length of one */
+        public void setClass(String className) {
+            if (classes.length == 1) {
+                classes[0] = className;
+                return;
+            }
+
+            classes = new String[]{className};
+
+        }
+
+        public String[] getPackages() {
+            return packages;
+        }
+
+        public void setPackages(String[] packs) {
+            packages = packs;
+        }
+
+        public String[] getClasses() {
+            return classes;
+        }
+
+
+        public String getOperation() {
+            return operation;
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder sb = new StringBuilder();
+
+            if (packages != null) {
+                int last = packages.length - 1;
+                for (int i = 0; i < last; i++) {
+                    sb.append(packages[i]);
+                    sb.append('.');
+                }
+
+                sb.append(packages[last]);
+            }
+            sb.append('.'); // this dot is there even if package is null
+
+            int last = classes.length - 1;
+            for (int i = 0; i < last; i++) {
+                sb.append(classes[i]);
+                sb.append('$');
+            }
+
+            sb.append(classes[last]);
+
+            if (operation != null) {
+                sb.append('#');
+                sb.append(operation);
+            }
+
+            return sb.toString();
+        }
+
+        @Override
+        public boolean equals(Object o) {
```

Comment:
this equals and hashcode look unusual. I'd just start with letting it generate by your IDE.

There is also the general contract, that "equal objects must have equal hash codes" (see https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#equals-java.lang.Object-)

You'll first need to decide, when two QualifiedName instances can be considered equal: Is it enough, if operation (method) names are the same, or do the class names need to match as well, ...

Looking at the toString method, I guess, that everything needs to be equal. So, when generating equals/hashCode in the IDE, just use all the fields, and you should be fine.


The problem of having a non-matching equals/hashCode implementation is, that you might put an instance into a HashMap, but it could happen, that "contains" or "get" will never find it again... hard to debug. And it even depends on the JVM implementation (and behavior could change between java updates).